miner/worker.go

```diff
@@ -360,619 +238,200 @@ func (w *worker) createTxCmp() func(tx1 *types.Transaction, tx2 *types.Transacti
-	}
-
-	txFeeRecipient := w.txFeeRecipient
-	if !w.chainConfig.IsDonut(header.Number) && w.txFeeRecipient != w.validator {
-		txFeeRecipient = w.validator
-		log.Warn("TxFeeRecipient and Validator flags set before split etherbase fork is active. Defaulting to the given validator address for the coinbase.")
-	}
-
-	// Only set the coinbase if our consensus engine is running (avoid spurious block rewards)
-	if w.isRunning() {
-		if txFeeRecipient == (common.Address{}) {
-			log.Error("Refusing to mine without etherbase")
-			return
-		}
-		header.Coinbase = txFeeRecipient
-	}
-	if err := w.engine.Prepare(w.chain, header); err != nil {
-		log.Error("Failed to prepare header for mining", "err", err)
-		return
-	}
-	// Start record block construction time after `engine.Prepare` to exclude the sleep time
-	defer func(start time.Time) { w.blockConstructGauge.Update(time.Since(start).Nanoseconds()) }(time.Now())
-
-	// If we are care about TheDAO hard-fork check whether to override the extra-data or not
-	if daoBlock := w.chainConfig.DAOForkBlock; daoBlock != nil {
-		// Check whether the block is among the fork extra-override range
-		limit := new(big.Int).Add(daoBlock, params.DAOForkExtraRange)
-		if header.Number.Cmp(daoBlock) >= 0 && header.Number.Cmp(limit) < 0 {
-			// Depending whether we support or oppose the fork, override differently
-			if w.chainConfig.DAOForkSupport {
-				header.Extra = common.CopyBytes(params.DAOForkBlockExtra)
-			} else if bytes.Equal(header.Extra, params.DAOForkBlockExtra) {
-				header.Extra = []byte{} // If miner opposes, don't let it use the reserved extra-data
+		if w.isRunning() {
+			if h, ok := w.engine.(consensus.Handler); ok {
+				h.NewWork()
 			}
-		}
-	}
-	// Could potentially happen if starting to mine in an odd state.
-	err := w.makeCurrent(parent, header)
-	if err != nil {
-		log.Error("Failed to create mining context", "err", err)
-		return
-	}
-	// Create the current work task and check any fork transitions needed
-	env := w.current
-	if w.chainConfig.DAOForkSupport && w.chainConfig.DAOForkBlock != nil && w.chainConfig.DAOForkBlock.Cmp(header.Number) == 0 {
-		misc.ApplyDAOHardFork(env.state)
-	}
-
-	if !noempty && !w.isIstanbulEngine() && atomic.LoadUint32(&w.noempty) == 0 {
-		// Create an empty block based on temporary copied state for sealing in advance without waiting block
-		// execution finished.
-		w.commit(nil, false, tstart)
-	}
 
-	istanbulEmptyBlockCommit := func() {
-		if !noempty && w.isIstanbulEngine() {
-			w.commit(nil, false, tstart)
+			go w.constructAndSubmitNewBlock(taskCtx)
+		} else {
+			go w.constructPendingStateBlock(taskCtx, txsCh)
 		}
 	}
 
-	w.updateSnapshot()
-
-	// Play our part in generating the random beacon.
-	if w.isRunning() && random.IsRunning() {
-		istanbul, ok := w.engine.(consensus.Istanbul)
-		if !ok {
-			log.Crit("Istanbul consensus engine must be in use for the randomness beacon")
-		}
+	for {
+		select {
+		case <-w.startCh:
```

Comment:
if we call start() twice that means we will cancel current work and start again, feature or bug?

Reply:
That's correct. Unintended consequence that doesn't really matter. We want to cancel the pending block and start proposing blocks on `start`. It would be possible to gate sending to the start channel on the current miner state, but it doesn't seem like a big issue.